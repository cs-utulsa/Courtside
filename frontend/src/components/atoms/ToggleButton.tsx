import React, { FC, useState } from 'react';
import { Text, Pressable, StyleSheet } from 'react-native';

type ToggleButtonProps = {
    text: string;
};

export const ToggleButton: FC<ToggleButtonProps> = ({ text }) => {
    const [selected, setSelected] = useState<boolean>(false);

    return (
        <Pressable
            onPress={() => setSelected(!selected)}
            style={[
                styles.button,
                selected ? styles.btnSelected : styles.btnNotSelected,
            ]}
        >
            <Text
                style={[
                    styles.buttonText,
                    selected ? styles.selectedText : styles.unselectedText,
                ]}
            >
                {text}
            </Text>
        </Pressable>
    );
};

const styles = StyleSheet.create({
    button: {
        borderRadius: 5,
        paddingHorizontal: 8,
        paddingVertical: 10,
        marginHorizontal: 5,
        marginVertical: 5,
        borderWidth: 2,
        borderColor: '#EE6730',
    },
    buttonText: {
        fontSize: 14,
        fontWeight: '500',
    },
    btnSelected: {
        backgroundColor: '#EE6730',
    },
    btnNotSelected: {
        backgroundColor: 'white',
    },
    selectedText: {
        color: 'white',
    },
    unselectedText: {
        color: '#EE6730',
    },
});
