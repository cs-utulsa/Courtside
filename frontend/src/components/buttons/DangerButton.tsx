import React, { FC } from 'react';
import { Pressable, StyleSheet, Text } from 'react-native';

type DangerButtonProps = {
    /** The text describing the action of the button */
    text: string;
    /** Function defining what the button does when pressed */
    onPress: () => void;
};

/**
 * A button colored red with red text expressing that when pressed the user will do a dangerous activity
 */
export const DangerButton: FC<DangerButtonProps> = ({ text, onPress }) => {
    return (
        <Pressable onPress={onPress} style={styles.button}>
            <Text style={styles.text}>{text}</Text>
        </Pressable>
    );
};

const styles = StyleSheet.create({
    /** Styles for the actual button */
    button: {
        width: '85%',
        backgroundColor: '#ffcfcd',
        paddingVertical: 15,
        borderRadius: 10,
        marginVertical: 10,
    },
    /** Styles for the text of the button */
    text: {
        textAlign: 'center',
        color: 'red',
        fontSize: 16,
        textTransform: 'uppercase',
    },
});
