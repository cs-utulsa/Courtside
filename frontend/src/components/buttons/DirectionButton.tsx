import React, { FC } from 'react';
import { Pressable, Text, StyleSheet } from 'react-native';
import { MaterialIcons } from '@expo/vector-icons';

type DirectionButtonProps = {
    /** The direction that the arrow is facing */
    direction: 'left' | 'right';
    /** THe text displayed with the button */
    text: string;
    /** The function that is called when the button is pressed */
    onPress: () => void;
};

/**
 * A button that can display either a left arrow or a right arrow next to some text
 * @example
 * const direction = 'left';
 * const text = 'Going Left!';
 * const onPress = () => navigate("Screen to the left");
 * return <DirectionButton direction={direction} text={text} onPress={onPress} />
 */
export const DirectionButton: FC<DirectionButtonProps> = ({
    direction,
    text,
    onPress,
}) => {
    // specify the name of the icon to display
    const iconName = direction === 'left' ? 'chevron-left' : 'chevron-right';

    return (
        <Pressable
            style={[
                styles.button,
                direction === 'left' ? styles.buttonLeft : styles.buttonRight,
            ]}
            onPress={onPress}
        >
            <Text style={styles.text}>{text}</Text>
            <MaterialIcons name={iconName} color="#60DDF7" size={35} />
        </Pressable>
    );
};

const styles = StyleSheet.create({
    /** Styles for the button text */
    text: {
        color: '#60DDF7',
        fontSize: 25,
        marginEnd: -5,
    },
    /** Styles for the button */
    button: {
        display: 'flex',
        flexDirection: 'row',
        alignItems: 'center',
        marginRight: 15,
        marginBottom: 10,
    },
    /** Styles for when the button's direction is left */
    buttonLeft: {
        justifyContent: 'flex-start',
    },
    /** Styles for when the button's direction is right */
    buttonRight: {
        justifyContent: 'flex-end',
    },
});
