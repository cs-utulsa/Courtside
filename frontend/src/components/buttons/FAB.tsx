import React, { FC, ReactNode } from 'react';
import { Pressable, StyleSheet } from 'react-native';

type FABProps = {
    children: ReactNode;
    onPress: () => void;
    position?: 'left' | 'right';
    color: string;
};

export const FAB: FC<FABProps> = ({
    children,
    onPress,
    position = 'right',
    color,
}) => {
    let positionStyles = {};

    if (position === 'left') {
        positionStyles = {
            bottom: 15,
            left: 15,
        };
    } else if (position === 'right') {
        positionStyles = {
            bottom: 15,
            right: 15,
        };
    }

    const colorStyles = {
        backgroundColor: color,
    };

    return (
        <Pressable
            style={[styles.btn, positionStyles, colorStyles]}
            onPress={onPress}
            testID="FAB"
        >
            {children}
        </Pressable>
    );
};

const styles = StyleSheet.create({
    btn: {
        position: 'absolute',
        borderRadius: 10,
        width: 60,
        height: 60,
        alignItems: 'center',
        justifyContent: 'center',
    },
});
