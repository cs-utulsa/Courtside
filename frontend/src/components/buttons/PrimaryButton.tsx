import { ORANGE } from '@styles/colors';
import React, { FC } from 'react';
import { View, Pressable, Text, StyleSheet } from 'react-native';

type PrimaryButtonProps = {
    onPress: () => void;
};

export const PrimaryButton: FC<PrimaryButtonProps> = ({ onPress }) => {
    return (
        <View style={styles.container}>
            <Pressable style={styles.btn} onPress={onPress}>
                <Text style={styles.btnText}>Update</Text>
            </Pressable>
        </View>
    );
};

const styles = StyleSheet.create({
    container: {
        alignItems: 'center',
    },
    btn: {
        width: '90%',
        paddingVertical: 15,
        backgroundColor: 'white',
        borderRadius: 10,
        marginVertical: 10,
    },
    btnText: {
        textAlign: 'center',
        color: ORANGE,
        fontSize: 16,
    },
});
