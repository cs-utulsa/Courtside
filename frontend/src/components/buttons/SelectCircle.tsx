import React, { FC, useState } from 'react';
import { View, Image, StyleSheet, Pressable } from 'react-native';
import { ORANGE } from '../../styles/colors';

type SelectCircleProps = {
    /** Whether or not the button is selected initially */
    initialState?: boolean;
    /** Whether or not the user can change the state of the button. Defaults to false. */
    disabled?: boolean;
    /** The url of the image that should be displayed */
    url?: string;
    /** The size of the button in pixels */
    size: number;
    /** The function that runs when the status of the button changes. This must be defined if the button is not disabled */
    onSelectChanged?: (newStatus: boolean) => void;
};
/**
 * This component is a circle which displays an image. If it is not disabled, it can be selected on or off.
 * If disabled, this component is just a circle which displays an image
 * @example
 * const url = "/path/to/local/image";
 * const size = 100;
 * const onSelectChanged = (newStatus: boolean) => console.log("Is this button on?", newStatus);
 * return <SelectCircle url={url} size={size} onSelectChanged={onSelectChanged} />
 */
export const SelectCircle: FC<SelectCircleProps> = ({
    url,
    size,
    disabled = false,
    initialState = false,
    onSelectChanged = () => {},
}) => {
    const [selected, setSelected] = useState<boolean>(initialState);

    /**
     * Derives the size of the containing circle and the margins based on the size of the component
     * @returns an object with the height, width, borderRadius, and horizontal margin
     */
    const circleSize = () => {
        return {
            height: size - 6,
            width: size - 6,
            borderRadius: size / 2,
            marginHorizontal: 3,
        };
    };

    /**
     * Derive the size of the image from the size of the component
     * @returns an object with the height and width of the image
     * */
    const imgSize = () => {
        return { height: size * 0.6, width: size * 0.6 };
    };

    /**
     * Change the selection state when the button is pressed
     */
    const pressHandler = () => {
        setSelected(!selected);
        onSelectChanged(!selected);
    };

    return (
        <View
            style={[
                styles.circle,
                circleSize(),
                selected ? styles.selected : styles.notSelected,
            ]}
        >
            <Pressable onPress={pressHandler} disabled={disabled}>
                {url && (
                    <Image
                        source={{ uri: url }}
                        style={[styles.img, imgSize()]}
                    />
                )}
            </Pressable>
        </View>
    );
};

const styles = StyleSheet.create({
    /** The styles for the View surrounding the button */
    circle: {
        borderRadius: 40,
        borderWidth: 3,
        alignItems: 'center',
        justifyContent: 'center',
        backgroundColor: 'white',
    },
    /** Styles for the containing View when the button is in the not selected state */
    notSelected: {
        borderColor: '#808080',
    },
    /** Styles for the containing View when the button is in the selected state */
    selected: {
        borderColor: ORANGE,
    },
    /** Styles for the image that is displayed inside the button */
    img: {
        resizeMode: 'contain',
        width: 55,
        height: 55,
    },
});
