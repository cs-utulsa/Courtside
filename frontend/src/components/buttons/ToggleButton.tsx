import React, { FC, useState } from 'react';
import { Text, Pressable, StyleSheet } from 'react-native';
import { ORANGE } from '../../styles/colors';

type ToggleButtonProps = {
    /** Whether or not the button is toggled initially */
    initial: boolean;
    /** Text displayed in the button */
    text: string;
    /**
     * Function that runs when the button is toggled
     * @param {boolean} on that state the button is changing to (either on or off)
     */
    onToggle: (on: boolean) => void;
};

/**
 * A button that toggles between on and off when you press it.
 * @example
 * const initial={true}
 * const text="Button"
 * const onToggle = (on: boolean) => console.log("Is it on?", on);
 * return <ToggleButton initial={initial} text={text} onToggle={onToggle} />
 */
export const ToggleButton: FC<ToggleButtonProps> = ({
    initial,
    text,
    onToggle,
}) => {
    const [selected, setSelected] = useState<boolean>(initial);

    /**
     * toggles the button on or off when the button is pressed
     */
    const onPress = () => {
        const newValue = !selected;

        setSelected(newValue);
        onToggle(newValue);
    };

    return (
        <Pressable
            onPress={onPress}
            style={[
                styles.button,
                selected ? styles.btnSelected : styles.btnNotSelected,
            ]}
        >
            <Text
                style={[
                    styles.buttonText,
                    selected ? styles.selectedText : styles.unselectedText,
                ]}
            >
                {text}
            </Text>
        </Pressable>
    );
};

const styles = StyleSheet.create({
    /** Styles for the button in all states */
    button: {
        borderRadius: 5,
        paddingHorizontal: 8,
        paddingVertical: 10,
        marginHorizontal: 5,
        marginVertical: 5,
        borderWidth: 2,
        borderColor: ORANGE,
    },
    /** Styles for the button text in all states */
    buttonText: {
        fontSize: 14,
        fontWeight: '500',
    },
    /** Styles for the button when the button is selected */
    btnSelected: {
        backgroundColor: ORANGE,
    },
    /** Styles for the button when the button is  not selected */
    btnNotSelected: {
        backgroundColor: 'white',
    },
    /** Styles for the text when the button is selected */
    selectedText: {
        color: 'white',
    },
    /** Styles for the text when the button is not selected */
    unselectedText: {
        color: ORANGE,
    },
});
