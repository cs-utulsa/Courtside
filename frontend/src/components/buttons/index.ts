export { LeftButton } from './LeftButton';
export { RightButton } from './RightButton';
export { ToggleButton } from './ToggleButton';
export { SelectCircle } from './SelectCircle';
export { PrimaryButton } from './PrimaryButton';
