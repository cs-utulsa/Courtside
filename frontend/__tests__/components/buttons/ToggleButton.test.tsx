/* jshint ignore:start */
import React from 'react';
import { ToggleButton } from '@components/buttons/ToggleButton';
import { render, screen, fireEvent } from '@testing-library/react-native';
import '@testing-library/jest-native';

describe('Testing ToggleButton component', () => {
    it("shows as toggled when button's initial value is true", () => {
        const json = render(
            <ToggleButton initial={true} text="Text" onToggle={jest.fn()} />
        ).toJSON();

        expect(json).toMatchSnapshot();
    });

    it('shows as not toggled when button` initial value is false', () => {
        const json = render(
            <ToggleButton initial={false} text="Text" onToggle={jest.fn()} />
        ).toJSON();

        expect(json).toMatchSnapshot();
    });

    it('shows the given text', () => {
        render(
            <ToggleButton initial={false} text="Text" onToggle={jest.fn()} />
        );

        const btn = screen.getByText('Text');
        expect(btn).toBeOnTheScreen();
    });

    it('calls onToggle function when pressed', () => {
        const mockFn = jest.fn();

        render(<ToggleButton initial={false} text="Text" onToggle={mockFn} />);

        expect(mockFn).not.toHaveBeenCalled();

        const btn = screen.getByText('Text');

        fireEvent.press(btn);

        expect(mockFn).toHaveBeenCalledTimes(1);

        fireEvent.press(btn);
        expect(mockFn).toHaveBeenCalledTimes(2);
    });

    it('changes state when pressed', () => {
        render(
            <ToggleButton initial={false} text="Text" onToggle={jest.fn()} />
        );

        const btn = screen.getByText('Text');

        expect(screen.getByTestId('not-selected')).toBeOnTheScreen();

        fireEvent.press(btn);
        expect(screen.getByTestId('selected')).toBeOnTheScreen();

        fireEvent.press(btn);
        expect(screen.getByTestId('not-selected')).toBeOnTheScreen();
    });
});
/* jshint ignore:end */