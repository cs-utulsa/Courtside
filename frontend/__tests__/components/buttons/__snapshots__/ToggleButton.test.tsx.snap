// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing ToggleButton component shows as not toggled when button\` initial value is false 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "borderColor": "#EE6730",
        "borderRadius": 5,
        "borderWidth": 2,
        "marginHorizontal": 5,
        "marginVertical": 5,
        "paddingHorizontal": 8,
        "paddingVertical": 10,
      },
      Object {
        "backgroundColor": "white",
      },
    ]
  }
  testID="not-selected"
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 14,
          "fontWeight": "500",
        },
        Object {
          "color": "#EE6730",
        },
      ]
    }
  >
    Text
  </Text>
</View>
`;

exports[`Testing ToggleButton component shows as toggled when button's initial value is true 1`] = `
<View
  accessible={true}
  collapsable={false}
  focusable={true}
  onBlur={[Function]}
  onClick={[Function]}
  onFocus={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Array [
      Object {
        "borderColor": "#EE6730",
        "borderRadius": 5,
        "borderWidth": 2,
        "marginHorizontal": 5,
        "marginVertical": 5,
        "paddingHorizontal": 8,
        "paddingVertical": 10,
      },
      Object {
        "backgroundColor": "#EE6730",
      },
    ]
  }
  testID="selected"
>
  <Text
    style={
      Array [
        Object {
          "fontSize": 14,
          "fontWeight": "500",
        },
        Object {
          "color": "white",
        },
      ]
    }
  >
    Text
  </Text>
</View>
`;
